### ISSUE TITLE
Nullable reference types bias towards use of 'var'

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#23281]
**Version Used**: Preview 11152017

I have observed several cases where changing the return type of a member results in cascading updates through the code. In many cases, local variables declared to hold the result of invoking these members were already properly used, and the only reason why an update was required is the local variable was declared using the type name rather than using `var`.

It would be very helpful for the migration process if the semantics/warnings were slightly altered:

* When a nullable value is assigned to a non-nullable local variable in the initializer for the local variable, use a dedicated warning number for the situation and treat the variable as having been declared with a nullable type.
* Provide a code fix for this warning number which adds the missing `?`.

This warning can be disabled during the initial migration process for a project that prefers to not use `var`, with a final pass at the end to enable the warning and apply the code fix to clean up the code base.

**Comment by jcouv** [comment]
FYI @cston

**Comment by mavasani** [comment]
+1 on @sharwell's proposal

**Comment by jcouv** [comment]
I believe this is effectively what's implemented already, with a minor caveat.

In the following code, we use a dedicated error code for warning that affect intent (rather than safety). Those are called `W` warnings in the LDM notes, and they use `CS8600`.
Beyond the declaration, the declared type of the local doesn't really matter (we track it's nullability in the flow analysis anyways).
Then https://github.com/dotnet/roslyn/pull/26630 implements a fixer to make mark locals/parameters/etc as non-null. 

Once that fixer is checked in, I'll close the present issue as fixed.

```C#
public class C 
{
    public void M() 
    {
        string s = null; // warning CS8600: Converting null literal or possible null value to non-nullable type.
        s.ToString(); // here it doesn't matter that `s` was declared as nullable or not, we track null-state regardless
    }
}
```

**Comment by jcouv** [comment]
I think this was addressed. Closing.
